composition/src/index.js

```diff
@@ -0,0 +1,260 @@
+import { options } from 'preact';
+
+const EMPTY_STATE = {};
+
+/** @type {import('./internal').Component} */
+let currentComponent;
+
+let oldBeforeRender = options._render;
+options._render = vnode => {
+	if (oldBeforeRender) oldBeforeRender(vnode);
+
+	/** @type {import('./internal').Component} */
+	const c = (currentComponent = vnode._component);
+
+	/** the vnode component is a uninitialized composition component */
+	if (c.constructor.__compositions && !c.__compositions) {
+		c.__compositions = { u: [], w: [], e: [], x: {} };
+		const render = c.constructor(() => c.props);
+		c.constructor = props => render(props, c.__compositions.r);
+	}
+
+	/** the vnode component is a composition initialized component */
+	if (c.__compositions) {
+		// Update the ref to forward
+		if (vnode.ref) {
+			c.__compositions.r = vnode.ref;
+			vnode.ref = null;
+		}
+
+		// call all watch
+		c.__compositions.w.forEach(up => handleEffect(up, c));
+	}
+};
+
+let oldAfterDiff = options.diffed;
+options.diffed = vnode => {
+	if (oldAfterDiff) oldAfterDiff(vnode);
+
+	/** @type {import('./internal').Component} */
+	const c = vnode._component;
+	if (c && c.__compositions)
+		// handle all `effect`s
+		c.__compositions.e.forEach(up => handleEffect(up, c));
+};
+
+let oldBeforeUnmount = options.unmount;
+options.unmount = vnode => {
+	if (oldBeforeUnmount) oldBeforeUnmount(vnode);
+
+	/** @type {import('./internal').Component} */
+	const c = vnode._component;
+	if (c && c.__compositions) {
+		// cleanup `effect`s onCleanup
+		c.__compositions.e.forEach(cleanupEffect);
+		// call all onUnmounted lifecycle callbacks
+		c.__compositions.u.forEach(f => f());
+	}
+};
+
+const $Reactive = Symbol('reactive');
+
+export function createComponent(comp) {
+	comp.__compositions = true;
+	return comp;
+}
+
+export function watch(src, cb, dv) {
+	const vr = { value: dv };
+	Object.defineProperty(vr, $Reactive, {
+		get() {
+			return this.value;
+		}
+	});
+	const up = { src, cb, vr };
+	handleEffect(up, currentComponent);
+	currentComponent.__compositions.w.push(up);
+	return vr;
+}
+
+export function effect(src, cb) {
+	currentComponent.__compositions.e.push({ src, cb });
+}
+
+export function onMounted(cb) {
+	currentComponent._renderCallbacks.push(cb);
+}
+
+export function onUnmounted(cb) {
+	currentComponent.__compositions.u.push(cb);
+}
+
+export function provide(name, _value) {
+	const c = currentComponent;
+	if (!c.__compositions.c) {
+		c.__compositions.c = {};
+		c.getChildContext = () => c.__compositions.c;
+	}
+	c.__compositions.c[`__sC_${name}`] = { _component: c, _value };
+}
+
+export function inject(name, defaultValue) {
+	const c = currentComponent;
+
+	const ctx = c.context[`__sC_${name}`];
```

Comment:
you only seem to use `c` once I'd inline `currentComponent` here

Reply:
It's also used on line [112](https://github.com/preactjs/preact/pull/1923/files#diff-23369f01c5ea98c3654a47ca1abd2de7R112) to update the component when the injected variable is a value or reactive